cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -48,10 +48,10 @@ private class StdIterator extends Class {
 }
 
 /**
- * A type which can be used as an iterator
+ * Implements `Iterator`.
  */
-class Iterator extends Type {
-  Iterator() {
+private class IteratorImpl extends Iterator {
+  IteratorImpl() {
     this instanceof IteratorByTypedefs or
     exists(IteratorTraits it | it.getIteratorType() = this) or
     this instanceof StdIterator
```

Comment:
We don't really need `IteratorImpl`. `IteratorByTypedefs` and `StdIterator` can extend `Iterator` now that it's abstract, and the `IteratorTraits` case can be the charpred of a new class.

Reply:
Done.  Thanks for this suggestion, this is a much better design.